[Nano] iP

Nano-233 opened the pull request:
# Luna Task Manager 🚀  
**A sleek CLI app to organize your tasks without breaking a sweat!**  

## ✨ Key Features  
1. **Task Types**:  
  - `todo Bake a cake` → Simple tasks, no frills.  
  - `deadline Submit report /by 2025/05/30` → Add deadlines. ⏰  
  - `event F1 Singapore /from 2025/02/03 /to 2025/02/10` → Schedule events.  

2. **Task Control**:  
  - *Mark* tasks as done with **`mark [number]`** or *unmark* them with **`unmark [number]`**.  
  - Say goodbye to ~~clutter~~: **`delete [number]`** removes a task, or **`delete all`** nukes everything. 💥  

3. **Search & Organize**:  
  - Use **`list`** to see all tasks.  
  - **`find exams`** hunts for tasks with keyword "exams". 🔎  

## 🖥️ Main entry
```java
//Creates an instance of Luna and starts the application
public static void main(String[] args) {
        new Luna("data/Luna.txt").run();
}
```
## ✅ What’s Done?  
- [x] Add/delete tasks  
- [x] Mark/unmark system  
- [x] Auto-save to `data/Luna.txt`  
- [x] Search by keyword  

> [!TIP]
> Luna **auto-saves** your tasks, so you won’t lose progress even if you close the app!  

**Why Luna?** It’s *lightning-fast* ⚡, keeps your workflow **`/by`**-the-book, and makes task management *almost fun*.  

🔗 **Learn Java OOP principles**: [Oracle Java Docs](https://docs.oracle.com/javase/tutorial/)

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Nano-233 committed:
Renamed chatbot to Luna and added skeletal code

- Added greeting and exit message and ASCII art for Luna.

Nano-233 committed:
Added basic input-output for Luna chatbot

- Echos user input
- Loop to accept user input until "bye".

Nano-233 committed:
Added task management feature to Luna chatbot

- Users can add tasks to a list.
- Implemented "list" command to display all tasks.

Nano-233 committed:
Added functionality to mark and unmark tasks in Luna chatbot

- Implemented the "mark <task_number>" command to mark tasks as done.
- Implemented the "unmark <task_number>" command to mark tasks as not done.
- Enhanced the task list to display completion status with [X] or [ ].
- Ensured input validation for task numbers.

Nano-233 committed:
Refactored Luna chatbot to use a Task class for better structure

- Replaced separate arrays for tasks and statuses with a Task class.
- Added `mark()` and `unmark()` methods in the Task class to manage task states.
- Simplified task handling logic using the Task class.

Nano-233 committed:
Added support for Todo, Deadline, and Event tasks in Luna chatbot

- Implemented Todo, Deadline, and Event classes which extends from the Task class.
- Updated list command to display task types and relevant details.

Nano-233 committed:
Added Automated Text UI Testing For Luna

- Implemented a script to automate Text UI testing for the Luna chatbot.
- Adjusted ASCII art for compatibility across platforms.
- Improved test script to be more intuitive.

Nano-233 committed:
Added comprehensive error handling and updated Text UI testing

- Added error handling for invalid or partially correct commands.
- Updated the test cases to fit the new error handling system.

Nano-233 committed:
Added the delete command to Luna

- Added the ability to delete a task.
- Now uses arraylist instead of normal array.

Nano-233 committed:
Added ErrorType enum and updated test cases

- Introduced ErrorType enum to streamline error handling and improve code readability.
- Updated input and expected test cases to include scenarios for `delete` functionality and error handling.
- Ensured all commands (e.g., `todo`, `deadline`, `event`, `mark`, `unmark`, `delete`) produce consistent error messages for invalid inputs or formats.

Nano-233 committed:
Add file read/write functionality for task persistence

- Save tasks to `data/Luna.txt` whenever changes occur
- Load tasks from file on startup to maintain persistence
- Handle missing or corrupted files gracefully

Nano-233 committed:
Merge branch 'branch-Level-7'

Nano-233 committed:
Add date-time handling and validation for deadlines and events

- Implement `LocalDate` for proper date parsing and formatting.
- Update `Deadline` and `Event` classes to use date objects instead of raw strings
- Ensure input validation for date formats, rejecting incorrectly formatted dates
- Improve error handling to gracefully handle invalid or corrupted date inputs
- Update file storage format to correctly save and load dates

Nano-233 committed:
Merge branch 'branch-Level-8'

Nano-233 committed:
Refactor code to improve OOP structure

Luna.java very messy. Minimal OOP practice

Commands have common behaviours, classes are hard reads

Let's refactor code to improve oop structure and readability:
- Separate classes for each command inheriting from a common superclass
- Use classes to execute behaviours where possible

OOP is preferable here to ensure good maintainability, readability and
exetensibility

Nano-233 committed:
Merge branch 'branch-A-MoreOOP'

Nano-233 committed:
Refactor project structure by organizing classes into packages

All classes are in a single `java` directory, hard to manage

Let's group related classes into structured packages:
- `luna.command` for command-related classes
- `luna.storage` for file handling
- `luna.task` for task-related classes
- `luna.ui` for UI interactions
- `luna` for the main logic classes
- Add delete all command for convenience

This improves code maintainability, modularity, and scalability

Test files and paths are updated to reflect the new structure.

Nano-233 committed:
Merge branch 'branch-A-Packages'

Nano-233 committed:
Merge branch 'add-gradle-support'

Nano-233 committed:
Integrate gradle

Repetitive tasks present such as testing and building

Let's implement gradle:
- Automate the building and testing processes
- Check coding style

Using gradle ensures increased efficiency and automation

Nano-233 committed:
Merge branch 'branch-A-Gradle'

Nano-233 committed:
Add JUnit tests

Manual testing done by hand

Let's use JUnit to automate the testing process

Using JUnit allows for simpler streamlined testing

Nano-233 committed:
Merge branch 'branch-A-JUnit'

Nano-233 committed:
Add Javadoc comments to improve code documentation

The code is lacking Javadoc comments, making it more confusing

Add Javadoc comments to all non-private classes and methods

This improves maintainability and readability by providing structured documentation for each method and class.

Nano-233 committed:
Improve code to match coding standards

Some parts of the code do not fully adhere to the required coding standards

Let's refactor code to follow proper indentation, line wrapping, etc

Ensures better readability, maintainability, and compliance
with standard coding practices.

Nano-233 committed:
Add find command

No quick way of finding specific tasks if too many

Let's:
- Add a command to find tasks using a keyword
- Add a method to get task description for easier search

These changes allows for quick searches to find the desired task(s).

Nano-233 committed:
Merge branch 'branch-A-JavaDoc'

Nano-233 committed:
Merge branch-A-CodingStandard into master with resolved conflicts

Nano-233 committed:
Merge branch-Level-9 into master with resolved conflicts

Nano-233 committed:
Add checkstyle comment

Nano-233 committed:
Merge branch 'branch-A-Checkstyle'

gekjunxu reviewed (commented):
LGTM! 👍

Nano-233 committed:
Refactor CLI into GUI using JavaFX

Luna's interaction is handled entirely through the command line,
which limits user experience and usability.
The user must manually enter commands without visual cues,
and responses are printed sequentially in the terminal.

A graphical user interface (GUI) enhances usability by providing
a structured interface with a text box for input and a scrollable
dialog container for responses.
This improves readability, accessibility, and overall user engagement.

Let's:
- Integrate JavaFX into Luna to create a GUI
- Use class controllers to handle user input and display
- Use string returns instead of print statements

Using JavaFX's FXML structure keeps UI logic separate from chatbot logic,
improving maintainability.

Nano-233 committed:
Merge branch 'branch-Level-10'

Nano-233 committed:
Improve find command

Find task command is missing error handling

Let's:
- Standardize error handling to conform with other commands
- Add varargs support to search for multiple keywords at once

Standardized error handling will also provide example format
to the users as a guideline.

Nano-233 committed:
Merge branch 'branch-A-Varargs'

wj200 commented on src/main/java/luna/LunaException.java:
```diff
@@ -1,22 +1,57 @@
+         * @param specificMessage Additional information regarding the specific error.
+         * @return A detailed error message including the general and specific error descriptions.
+         */
         public String getMessage(String specificMessage) {
```
I like the use of camelCasing and naming methods as an action phrase/verb, which follows the naming conventions outlined in this course. This makes the code more readable and definitely improves code quality

wj200 commented on src/main/java/luna/Luna.java:
```diff
@@ -5,13 +5,37 @@
 import luna.ui.Ui;
 
-
+/**
```
Great use of comments outlining the purposes of each variable, method and class

wj200 commented on src/main/java/luna/LunaException.java:
```diff
@@ -1,22 +1,57 @@
 package luna;
```
header comment for the package is missing. It should typically include a clear overview of what the package is. More specifically a brief description of its purpose, the author/organization, the date created, and any relevant license information.

wj200 commented on src/main/java/luna/LunaException.java:
```diff
@@ -1,22 +1,57 @@
+    /**
+     * Defines various types of errors that can occur within Luna.
+     */
     public enum ErrorType {
```
I like the use of enumerations to restrict the possible valid ErrorType.

wj200 reviewed (commented):
LGTM! I also noticed the effort to include javadocs. Keep up the great effort

Nano-233 committed:
Ensure key assumptions hold with assertions

Luna assume that `storage` and `tasks` are always initialized and
that tasks added to `TaskList` are valid.

Error handling manages user input errors and file-related exceptions
but it does not enforce these core assumptions.

Let's add assertions to ensure that these assumptions hold.

These assertions verify invariants and indicates possible bugs.

Nano-233 committed:
Refactor task deletion into `TaskList` and clean up `AddCommand`

Task deletion logic reside in `DeleteCommand` instead of `TaskList`,
which is inconsistent as it currently encapsulates `AddCommand`.
`AddCommand` contain deep nesting.

The inconsistency reduce code quality and maintainability.
Deep nesting and repetitive validation logic reduces readability.

Let's:
- Move the deletion logic into `TaskList` to ensure both adding and
deleting tasks are handled within the same class.
- Let's refactor `AddCommand` by introducing helper methods
to validate inputs, reducing deep nesting and improving readability.

This improves cohesion in `TaskList`, makes command classes cleaner,
and enhances maintainability by following a consistent structure
for task management.

Nano-233 committed:
Refactor `loadTasks`, `printList`, and `findTasks` to use streams

`loadTasks`, `printList`, and `findTasks` use explicit loops and index,
making them less concise and harder to read.

Manually iterating adds nesting, and complicates logic.

Let's refactor these methods to use streams:
- `loadTasks` to simplify task loading.
- `printList` to format task lists concisely.
- `findTasks` to improve search readability.

Using streams enhance readability, reduce nesting,
and make task-related operations more maintainable.

Nano-233 committed:
Merge pull request #3 from Nano-233/branch-A-Assertions

Ensure key assumptions hold with assertions

Nano-233 committed:
Merge branch 'master' into branch-A-CodeQuality

Nano-233 committed:
Merge branch 'master' into branch-A-Streams

Nano-233 committed:
Merge pull request #1 from Nano-233/branch-A-CodeQuality

Refactor task deletion into `TaskList` and clean up `AddCommand`

Nano-233 committed:
Merge branch 'master' into branch-A-Streams

Nano-233 committed:
Merge pull request #2 from Nano-233/branch-A-Streams

Refactor `loadTasks`, `printList`, and `findTasks` to use streams

Nano-233 committed:
Set up CI

Nano-233 committed:
Fix styling issues of JUnit

Nano-233 committed:
Add `TagCommand` to Luna

Tasks have no tags to visually categorize them easily.

Larger number of tasks may lead to difficulty using the bot.

Let's:
- Add `TagCommand` class to support tagging of tasks.
- modify the `toFileFormat` method in task classes to include tags.
- Update `fromFileFormat` to support loading tasks with tags.
- Refactor `fromFileFormat` to improve readability and code quality.

Adding support of tagging by introducing `TagCommand` ensures
consistency of commands and user usage, and refactoring
`fromFileFormat` into different methods improves readability
and maintainability by reducing the complexity.
`ToFileFormat`ensures tags are saved in the third field
to ensure consistency regardless of the task type.

Nano-233 committed:
Update JUnit test cases

Nano-233 committed:
Merge pull request #4 from Nano-233/branch-C-Tagging

Add `TagCommand` to Luna

Nano-233 committed:
Fix entry point for gradle

Nano-233 committed:
Improve JavaFX GUI

The GUI lacks visual refinement,
making it appear plain and less engaging.
Messages do not have clear boundaries, and spacing is inconsistent.

A cleaner GUI improves readability, enhances user experience,
and makes interactions feel more polished.

Let's:
- Add padding and spacing to UI components for a
less cluttered appearance.
- Implement rounded dialog boxes with borders to
distinguish messages.
- Apply a modern color scheme for improved contrast between
user and bot messages.
- Ensure consistent text and button styling for better readability.
- Adjust ScrollPane and TextField aesthetics to align with the new UI.
- Ensure window scaling is supported.

These improvements makes the chatbot visually appealing while maintaining
clarity and usability.

Nano-233 committed:
Merge branch 'branch-A-BetterGui'

Nano-233 committed:
Fix naming convention of boolean

Nano-233 committed:
Enhance chatbot responses

Luna's responses are functional but lack personality in some commands.
This makes interactions feel robotic and less engaging.

Let's:
Add punctuation to create excitement.
Add emoticons such as `:D` to add emotions.

A chatbot with a distinct personality enhances user engagement
and retention, making conversations feel more human and lively.

Nano-233 committed:
Merge branch 'branch-A-Personality'

Nano-233 committed:
Fix line wrapping of Dialogboxes

Nano-233 committed:
Merge branch 'branch-A-BetterGui'

Nano-233 committed:
Refactor AddCommand to improve error handling and input validation

Input of `todo`, `deadline`, or `event` by user without further
arguments give the same error message.

This leads to users being unsure of the exact input format of
the specific task type.

Let's modify:
- Refactor AddCommand to add additional information to the generic
error message depending on the current task type.

Now, tasks with missing descriptions will provide a specific format
error message.

Nano-233 committed:
Merge branch 'branch-A-MoreErrorHandling'

Nano-233 committed:
Add more JUnit tests

Nano-233 committed:
Merge branch 'branch-A-MoreTesting'

Nano-233 committed:
Add documentation to JUnit tests

Nano-233 committed:
Merge branch 'branch-A-AiAssisted'

Nano-233 committed:
Add AI.md

Nano-233 committed:
Merge branch 'branch-A-AiAssisted'

Nano-233 committed:
Add application icon

Nano-233 committed:
Add user guide and screenshot

Nano-233 committed:
Merge branch 'branch-A-UserGuide'

Nano-233 committed:
Improve user guide

Nano-233 committed:
Merge branch 'branch-A-UserGuide'

Nano-233 committed:
Add exit guidance

Nano-233 committed:
Merge branch 'branch-A-UserGuide'

Nano-233 committed:
new file

Nano-233 committed:
Update new.txt

Nano-233 committed:
deleted

Nano-233 committed:
Update new.txt

Nano-233 committed:
branch changes

Nano-233 committed:
fixing merge conflicts

Nano-233 committed:
added line
